core/src/Dscp/Core/Foundation/Instances.hs

```diff
@@ -18,51 +18,263 @@ import Dscp.Util
+    encode a = encodeListLen 1
+            <> encodeWord (case a of
+                ATGAdded   -> 0
+                ATGRemoved -> 1
+                )
+    decode = do
+        decodeListLenOf 1
+        decodeWord >>= \case
+            0 -> return ATGAdded
+            1 -> return ATGRemoved
+            _ -> fail "Unexpected tag"
+
+instance Serialise ATGDelta where
+    encode = encode . getATGDelta
+    decode = ATGDelta <$> decode
+
+instance Serialise ATG where
+    encode = encode . getATGRoots
+    decode = ATG <$> decode
+
+instance Serialise Assignment where
+    encode (Assignment {..}) = mconcat
+        [ encodeListLen 4
+        , encode _aCourseId
+        , encode _aContentsHash
+        , encode _aType
+        , encode _aDesc
+        ]
+
+    decode = do
+        decodeListLenOf 4
+        _aCourseId     <- decode
+        _aContentsHash <- decode
+        _aType         <- decode
+        _aDesc         <- decode
+        return $ Assignment {..}
 
-instance Serialise Grade
-instance Serialise ATGNode
-instance Serialise ATGEdge
+instance Serialise AssignmentType where
+    encode a = encodeListLen 1
+            <> encodeWord (case a of
+                Regular     -> 0
+                CourseFinal -> 1
+                )
+    decode = do
+        decodeListLenOf 1
+        decodeWord >>= \case
+            0 -> return Regular
+            1 -> return CourseFinal
+            _ -> fail "Unexpected tag"
+
+instance Serialise SubmissionWitness where
+    encode (SubmissionWitness {..}) = mconcat
+        [ encodeListLen 2
+        , encode _swKey
+        , encode _swSig
+        ]
+
+    decode = do
+        decodeListLenOf 2
+        _swKey <- decode
+        _swSig <- decode
+        return $ SubmissionWitness {..}
+
+instance Serialise SignedSubmission where
+    encode (SignedSubmission {..}) = mconcat
+        [ encodeListLen 2
+        , encode _ssSubmission
+        , encode _ssWitness
+        ]
 
-instance Serialise ATGSubjectChange
-deriving instance Serialise ATGDelta
-deriving instance Serialise ATG
+    decode = do
+        decodeListLenOf 2
+        _ssSubmission <- decode
+        _ssWitness    <- decode
+        return $ SignedSubmission {..}
 
-instance Serialise Assignment
-instance Serialise AssignmentType
-instance Serialise SubmissionWitness
-instance Serialise SignedSubmission
-instance Serialise (DocumentType a)
+instance Serialise (DocumentType a) where
+    encode a = encodeListLen 1
+            <> encodeWord (case a of
+                Online  -> 0
+                Offline -> 1
+                )
+    decode = do
+        decodeListLenOf 1
+        decodeWord >>= \case
+            0 -> return Online
+            1 -> return Offline
+            _ -> fail "Unexpected tag"
 
 instance Serialise Day where
     encode = encode . toModifiedJulianDay
     decode = ModifiedJulianDay <$> decode
 
-instance Serialise EducationForm
-instance Serialise CertificateMeta
+-- consider that it is an Enum
```

Comment:
Just remembered that I left this comment behind, I'll remove it in the next commit